python/ray/dag/compiled_dag_node.py

```diff
@@ -26,119 +27,180 @@
 
 @DeveloperAPI
 def do_allocate_channel(
-    self, readers: List[Optional["ray.actor.ActorHandle"]], buffer_size_bytes: int
+    self,
+    readers: List[Optional["ray.actor.ActorHandle"]],
+    buffer_size_bytes: int,
+    reader_node_id: Optional["ray.NodeID"] = None,
 ) -> Channel:
     """Generic actor method to allocate an output channel.
 
     Args:
+        readers: The actor handles of the readers.
         buffer_size_bytes: The maximum size of messages in the channel.
-        num_readers: The number of readers per message.
+        reader_node_id: The node ID of the readers (must be the same for all readers).
 
     Returns:
         The allocated channel.
     """
-    self._output_channel = Channel(
-        readers,
-        buffer_size_bytes,
-    )
-    return self._output_channel
-
-
-def _wrap_exception(exc):
-    backtrace = ray._private.utils.format_error_message(
-        "".join(traceback.format_exception(type(exc), exc, exc.__traceback__)),
-        task_exception=True,
-    )
-    wrapped = RayTaskError(
-        function_name="do_exec_compiled_task",
-        traceback_str=backtrace,
-        cause=exc,
-    )
-    return wrapped
+    # We need to pass in `reader_node_id` to avoid a potential deadlock.
+    # Specifically, without predetermined `reader_node_id`, during channel creation,
+    # it makes a `ray.get()` blocking call to get the node id of the `readers`, which
+    # executes on the reader actor; and if the reader actor is the current actor which
+    # is creating the channel, there will be a deadlock.
```

Comment:
Ah, it should be possible to check that the current actor is not in the readers. You can check if `ray.get_runtime_context().current_actor`'s actor_id matches any in the `readers`.

Reply:
Done.